src/ring_theory/unique_factorization_domain.lean

```diff
@@ -292,49 +337,198 @@ def of_unique_irreducible_factorization {α : Type*} [integral_domain α]
-                (o.irreducible_factors ha0 _)
-                (o.irreducible_factors hb0 _)) $
-              calc multiset.prod (p :: o.factors x)
+        begin
+          cases eif x hx0 with fx hfx,
+          cases eif a ha0 with fa hfa,
+          cases eif b hb0 with fb hfb,
+          have h : multiset.rel associated (p :: fx) (fa + fb),
+          { apply uif,
+            { exact λ i hi, (multiset.mem_cons.1 hi).elim (λ hip, hip.symm ▸ hpi) (hfx.1 _), },
+            { exact λ i hi, (multiset.mem_add.1 hi).elim (hfa.1 _) (hfb.1 _), },
+            calc multiset.prod (p :: fx)
                   ~ᵤ a * b : by rw [hx, multiset.prod_cons];
                     exact associated_mul_mul (by refl)
-                      (o.factors_prod hx0)
-              ... ~ᵤ (o.factors a).prod * (o.factors b).prod :
-                associated_mul_mul
-                  (o.factors_prod ha0).symm
-                  (o.factors_prod hb0).symm
-              ... = _ : by rw multiset.prod_add,
-        let ⟨q, hqf, hq⟩ := multiset.exists_mem_of_rel_of_mem this
+                      (hfx.2.symm)
+              ... ~ᵤ (fa).prod * (fb).prod :
+                associated_mul_mul hfa.2 hfb.2
+              ... = _ : by rw multiset.prod_add, },
+          exact let ⟨q, hqf, hq⟩ := multiset.exists_mem_of_rel_of_mem h
           (multiset.mem_cons_self p _) in
         (multiset.mem_add.1 hqf).elim
           (λ hqa, or.inl $ (dvd_iff_dvd_of_rel_left hq).2 $
-            (dvd_iff_dvd_of_rel_right (o.factors_prod ha0)).1
+            (dvd_iff_dvd_of_rel_right hfa.2.symm).1
               (multiset.dvd_prod hqa))
           (λ hqb, or.inr $ (dvd_iff_dvd_of_rel_left hq).2 $
-            (dvd_iff_dvd_of_rel_right (o.factors_prod hb0)).1
-              (multiset.dvd_prod hqb))⟩,
-  ..o }
+            (dvd_iff_dvd_of_rel_right hfb.2.symm).1
+              (multiset.dvd_prod hqb))
+        end⟩, irreducible_of_prime⟩
+
+theorem unique_factorization_monoid.of_exists_unique_irreducible_factors
+  [comm_cancel_monoid_with_zero α]
+  (eif : ∀ (a : α), a ≠ 0 → ∃ f : multiset α, (∀b ∈ f, irreducible b) ∧ a ~ᵤ f.prod)
+  (uif : ∀ (f g : multiset α),
+  (∀x∈f, irreducible x) → (∀x∈g, irreducible x) → f.prod ~ᵤ g.prod → multiset.rel associated f g) :
+  unique_factorization_monoid α :=
+unique_factorization_monoid_of_exists_prime_factors (by {
+  convert eif,
+  simp_rw irreducible_iff_prime_of_exists_unique_irreducible_factors eif uif,
+})
+
+namespace unique_factorization_monoid
+variables [comm_cancel_monoid_with_zero α] [decidable_eq α] [nontrivial α] [normalization_monoid α]
+  [unique_factorization_monoid α]
+
+/-- Noncomputably determines the multiset of prime factors. -/
+noncomputable def factors (a : α) : multiset α := if h : a = 0 then 0 else
+multiset.map normalize $ classical.some (unique_factorization_monoid.exists_prime_factors a h)
+
+theorem factors_prod {a : α} (ane0 : a ≠ 0) : associated (factors a).prod a :=
```

Comment:
The conclusion of `exists_factors` is of the form `a ~ᵤ (factors a).prod`, so for consistency it might be better to turn the conclusion of `exists_factors` or `factors_prod` around.